<!-- Trabalho Desenvolvido na Data Science Academy - www.datascienceacademy.com.br -->
# <font color='blue'>Data Science Academy</font>
# <font color='blue'>Fundamentos de Linguagem Python - Do Básico a Aplicações de IA</font>
# <font color='blue'>Lista 3 de Exercícios</font>

Classificamos os 10 exercícios nas seguintes categorias:

- (Nível Baby)
- (Nível Aprendiz)
- (Nível Iniciante)
- (Nível Iniciante Plus)
- (Nível Pro)
- (Nível Master)
- (Nível Ninja)
- (Nível Ninja Pro Master)
- (Nível Ninja Pro Master das Galáxias)
- (Nível Ninja Pro Master das Galáxias Plus)

Divirta-se. :-)

In [1]:
# Importa o NumPy
import numpy as np

#### Exercício 1: Selecionando uma Coluna Específica (Nível Baby)

Dado a matriz 4x4 abaixo, crie um novo array contendo apenas os elementos da terceira coluna.

In [2]:
# Cria a matriz
matriz = np.array([
    [1, 2, 3, 4],
    [5, 6, 7, 8],
    [9, 10, 11, 12],
    [13, 14, 15, 16]
])

In [3]:
print(matriz)

[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]]


In [4]:
# Solução

# Para selecionar uma coluna, usamos ':' para as linhas e o índice da coluna desejada
# O ':' significa "todas as linhas"
# O índice 2 refere-se à terceira coluna (pois a indexação em Python começa em 0)
terceira_coluna = matriz[:, 2]

print("\nTerceira coluna da matriz:\n")
print(terceira_coluna)

# Saída esperada: [ 3  7 11 15]


Terceira coluna da matriz:

[ 3  7 11 15]


In [5]:
# CUIDADO.... ISSO AQUI GERA RESULTADO COMPLETAMENTE DIFERENTE:
matriz[:, 1:2]

array([[ 2],
       [ 6],
       [10],
       [14]])

In [6]:
# CUIDADO.... ISSO AQUI GERA RESULTADO COMPLETAMENTE DIFERENTE:
matriz[:, 2:2]

array([], shape=(4, 0), dtype=int64)

O resultado é vazio porque o fatiamento matriz[:, 2:2] não seleciona nenhuma coluna. Quando usamos o operador de slice em Python (inicio:fim), o índice fim é exclusivo, ou seja, ele não é incluído no resultado.

Em nosso caso:

- 2:2 significa “comece na coluna de índice 2 e vá até antes da coluna de índice 2”.

- Como o início e o fim são iguais, não há nenhuma coluna para pegar.

- O NumPy então retorna um array com 4 linhas e 0 colunas: array([], shape=(4, 0), dtype=int64), ou seja, vazio.

In [7]:
# Já isso aqui gera o resultado esperado!
matriz[:, 2:3]

array([[ 3],
       [ 7],
       [11],
       [15]])

#### Exercício 2: Extraindo um Bloco (Submatriz) (Nível Aprendiz)

A partir da mesma matriz do exercício 1, extraia o bloco central 2x2, que contém os números 6, 7, 10 e 11.

In [8]:
print(matriz)

[[ 1  2  3  4]
 [ 5  6  7  8]
 [ 9 10 11 12]
 [13 14 15 16]]


In [9]:
# Solução

# Extração dos dados da matriz
bloco_central = matriz[1:3, 1:3]

print("\nBloco central 2x2:\n")
print(bloco_central)

# Saída esperada:
# [[ 6  7]
#  [10 11]]


Bloco central 2x2:

[[ 6  7]
 [10 11]]


#### Exercício 3: Produto de Matrizes (Nível Iniciante)

Dadas as duas matrizes A e B abaixo, calcule o produto matricial entre A e B.

In [10]:
# Cria as matrizes
A = np.array([[1, 2, 3], [4, 5, 6]])       # Matriz 2x3
B = np.array([[7, 8], [9, 10], [11, 12]])  # Matriz 3x2

In [11]:
# Solução

# O operador @ é a forma recomendada para o produto de matrizes em Python 3.5+
C = A @ B

print("Matriz A (2x3):")
print(A)
print("\nMatriz B (3x2):")
print(B)
print("\nProduto Matricial C (2x2):")
print(C)

# Saída esperada:
# [[ 58  64]
#  [139 154]]

Matriz A (2x3):
[[1 2 3]
 [4 5 6]]

Matriz B (3x2):
[[ 7  8]
 [ 9 10]
 [11 12]]

Produto Matricial C (2x2):
[[ 58  64]
 [139 154]]


Passo a Passo Data Science Academy:
<!-- Trabalho Desenvolvido na Data Science Academy - www.datascienceacademy.com.br -->
- C = A @ B: O operador @ realiza a multiplicação de matrizes. Ele calcula a soma dos produtos dos elementos das linhas de A pelos elementos das colunas de B. O número de colunas da primeira matriz (3) deve ser igual ao número de linhas da segunda matriz (3). O resultado é uma matriz com o número de linhas da primeira e o número de colunas da segunda (2x2).

#### Exercício 4: Selecionando Linhas Pares e Colunas Ímpares (Nível Iniciante Plus)

Dada a matriz 9x9 abaixo, crie um novo array que contenha apenas os valores das linhas de índice par e as colunas de índice ímpar.

In [12]:
# Cria a matriz
matriz = np.arange(81).reshape(9, 9)

In [13]:
print(matriz)

[[ 0  1  2  3  4  5  6  7  8]
 [ 9 10 11 12 13 14 15 16 17]
 [18 19 20 21 22 23 24 25 26]
 [27 28 29 30 31 32 33 34 35]
 [36 37 38 39 40 41 42 43 44]
 [45 46 47 48 49 50 51 52 53]
 [54 55 56 57 58 59 60 61 62]
 [63 64 65 66 67 68 69 70 71]
 [72 73 74 75 76 77 78 79 80]]


In [14]:
# Solução

# Usamos o passo no slicing (start:stop:step)
# Para as linhas: '::2' significa do início ao fim, pulando de 2 em 2 (0, 2, 4)
# Para as colunas: '1::2' significa do índice 1 ao fim, pulando de 2 em 2 (1, 3)
resultado = matriz[::2, 1::2]

print("Matriz Original:")
print(matriz)
print("\nLinhas Pares e Colunas Ímpares:")
print(resultado)

# Saída esperada:
# [[ 1  3  5  7]
#  [19 21 23 25]
#  [37 39 41 43]
#  [55 57 59 61]
#  [73 75 77 79]]

Matriz Original:
[[ 0  1  2  3  4  5  6  7  8]
 [ 9 10 11 12 13 14 15 16 17]
 [18 19 20 21 22 23 24 25 26]
 [27 28 29 30 31 32 33 34 35]
 [36 37 38 39 40 41 42 43 44]
 [45 46 47 48 49 50 51 52 53]
 [54 55 56 57 58 59 60 61 62]
 [63 64 65 66 67 68 69 70 71]
 [72 73 74 75 76 77 78 79 80]]

Linhas Pares e Colunas Ímpares:
[[ 1  3  5  7]
 [19 21 23 25]
 [37 39 41 43]
 [55 57 59 61]
 [73 75 77 79]]


#### Exercício 5: Somando Valor a Uma Submatriz (Nível Pro)

Dada a matriz abaixo (preenchida com zeros) adicione o valor 5 apenas ao bloco central 2x2.

In [15]:
# Cria a matriz
matriz = np.zeros((4, 4), dtype = int)

In [16]:
print(matriz)

[[0 0 0 0]
 [0 0 0 0]
 [0 0 0 0]
 [0 0 0 0]]


In [17]:
# Solução

# Primeiro, selecionamos o bloco central 2x2 usando slicing
# Em seguida, usamos o operador '+=' para somar 5 a cada elemento da seleção
matriz[1:3, 1:3] += 5

print("Matriz com o centro modificado:")
print(matriz)

# Saída esperada:
# [[0 0 0 0]
#  [0 5 5 0]
#  [0 5 5 0]
#  [0 0 0 0]]

Matriz com o centro modificado:
[[0 0 0 0]
 [0 5 5 0]
 [0 5 5 0]
 [0 0 0 0]]


#### Exercício 6: Normalização de Uma Matriz (Nível Master)

Normalize a matriz abaixo. 

A normalização (Z-score) é feita subtraindo a média de todos os elementos de cada elemento e, em seguida, dividindo pelo desvio padrão. 

Fórmula: (X - media) / desvio_padrao.

In [18]:
# Cria a matriz
matriz = np.array([[10, 20, 30], [40, 50, 60], [70, 80, 90]])

In [19]:
print(matriz)

[[10 20 30]
 [40 50 60]
 [70 80 90]]


In [20]:
# Solução

# 1. Calcular a média de todos os elementos da matriz
media = np.mean(matriz)

# 2. Calcular o desvio padrão de todos os elementos
desvio_padrao = np.std(matriz)

# 3. Aplicar a fórmula da normalização
matriz_normalizada = (matriz - media) / desvio_padrao

print("Média:", media)
print("Desvio Padrão:", f"{desvio_padrao:.2f}")
print("\nMatriz Normalizada:")
print(np.round(matriz_normalizada, 2))

Média: 50.0
Desvio Padrão: 25.82

Matriz Normalizada:
[[-1.55 -1.16 -0.77]
 [-0.39  0.    0.39]
 [ 0.77  1.16  1.55]]


Passo a Passo Data Science Academy:

- media = np.mean(matriz): Calcula a média de todos os números na matriz (que é 50).

- desvio_padrao = np.std(matriz): Calcula o desvio padrão (que é aprox. 25.82).

- matriz_normalizada = (matriz - media) / desvio_padrao: Aqui o NumPy usa broadcasting. Ele subtrai a media (um escalar) de cada elemento da matriz. Em seguida, divide cada elemento do resultado pelo desvio_padrao (outro escalar).

Aqui aplicamos o conceito de vetorização do NumPy (estudado no capítulo anterior).

<!-- Trabalho Desenvolvido na Data Science Academy - www.datascienceacademy.com.br -->

#### Exercício 7: Substituindo Valores com Base em Uma Condição (Nível Ninja)

Dada a matriz abaixo, crie uma nova matriz onde todos os números maiores que 8 sejam substituídos pelo valor -1. Crie uma cópia da matriz antes de fazer a operação.

In [21]:
# Cria a matriz
dados = np.arange(16).reshape(4, 4)

In [22]:
print(dados)

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]]


In [23]:
# Solução

# Copiamos a matriz para não alterar a original
matriz_modificada = dados.copy()

# Usamos indexação booleana para encontrar onde a condição é verdadeira
# E então atribuímos o novo valor (-1) a essas posições
matriz_modificada[matriz_modificada > 8] = -1

print("Matriz Original:")
print(dados)
print("\nMatriz Modificada:")
print(matriz_modificada)

# Saída esperada:
# [[ 0  1  2  3]
#  [ 4  5  6  7]
#  [ 8 -1 -1 -1]
#  [-1 -1 -1 -1]]

Matriz Original:
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]]

Matriz Modificada:
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8 -1 -1 -1]
 [-1 -1 -1 -1]]


Passo a Passo Data Science Academy:

- matriz_modificada = dados.copy(): É uma boa prática criar uma cópia para preservar o array original.

- matriz_modificada > 8: Esta operação cria uma matriz booleana (de True/False) com as mesmas dimensões da original, onde cada posição é True se o elemento correspondente for maior que 8, e False caso contrário.

- matriz_modificada[...] = -1: Usamos essa matriz booleana como um "filtro" ou "máscara". O valor -1 é atribuído a todas as posições da matriz_modificada onde a máscara booleana é True.

#### Exercício 8: Inversa de Uma Matriz (Nível Ninja Pro Master)

Calcule a matriz inversa da matriz A. Depois, verifique seu trabalho calculando o produto de A pela sua inversa (o resultado deve ser a matriz identidade).

A matriz inversa de uma matriz quadrada A é outra matriz, chamada A⁻¹, que quando multiplicada por A resulta na matriz identidade (uma matriz que tem 1 na diagonal principal e 0 nos outros elementos). Em outras palavras, A × A⁻¹ = I. A inversa só existe para matrizes quadradas que não sejam singulares, ou seja, que tenham determinante diferente de zero. Ela é usada para resolver sistemas lineares, desfazer transformações e em várias aplicações de álgebra linear.

O determinante é um número único calculado a partir dos elementos de uma matriz quadrada (mesmo número de linhas e colunas) que resume algumas propriedades fundamentais da matriz. Ele indica, por exemplo, se a matriz é invertível (determinante diferente de zero) ou singular (determinante igual a zero).

In [24]:
# Cria a matriz
A = np.array([[4, 7], [2, 6]])

In [25]:
print(A)

[[4 7]
 [2 6]]


In [26]:
# Solução

# A função inv() do submódulo linalg calcula a inversa
A_inversa = np.linalg.inv(A)

# Verificação: o produto de uma matriz pela sua inversa é a matriz identidade
identidade = A @ A_inversa

print("Matriz Original A:")
print(A)
print("\nMatriz Inversa de A:")
print(np.round(A_inversa, 2))
print("\nProduto A * A_inversa (deve ser a Matriz Identidade):")

# Usamos np.round() para lidar com pequenas imprecisões de ponto flutuante
# abs() retorna os valores absolutos
print(abs(np.round(identidade)))

Matriz Original A:
[[4 7]
 [2 6]]

Matriz Inversa de A:
[[ 0.6 -0.7]
 [-0.2  0.4]]

Produto A * A_inversa (deve ser a Matriz Identidade):
[[1. 0.]
 [0. 1.]]


Passo a Passo Data Science Academy:

- A_inversa = np.linalg.inv(A): O submódulo linalg do NumPy contém funções de álgebra linear. inv() é a função para calcular a inversa de uma matriz. Uma inversa só existe para matrizes quadradas cujo determinante não é zero.

- identidade = A @ A_inversa: Calculamos o produto matricial para verificação.

- np.round(...): Devido à forma como os computadores lidam com números de ponto flutuante, o resultado pode ter valores muito pequenos (como 1.11e-16) em vez de zero. np.round() arredonda esses valores para facilitar a visualização e confirmar que o resultado é, de fato, a matriz identidade.

#### Exercício 9: Resolvendo Um Sistema de Equações Lineares (Nível Ninja Pro Master das Galáxias)
<!-- Trabalho Desenvolvido na Data Science Academy - www.datascienceacademy.com.br -->
Resolva o seguinte sistema de equações lineares:

2x + y = 8

x + 3y = 7

Represente o sistema na forma matricial Ax = b e encontre o vetor x (que contém os valores de x e y).

In [27]:
# Solução

# Matriz A contém os coeficientes das variáveis (lado esquerdo da equação)
A = np.array([[2, 1], [1, 3]])

# Vetor b contém as constantes (lado direito da equação)
b = np.array([8, 7])

# A função solve() do submódulo linalg resolve o sistema Ax = b
x = np.linalg.solve(A, b)

print("Matriz de Coeficientes A:")
print(A)
print("\nVetor de Constantes b:")
print(b)
print(f"\nSolução (x, y): {np.round(x, 2)}") 

# Saída esperada: [3.4 1.2]

Matriz de Coeficientes A:
[[2 1]
 [1 3]]

Vetor de Constantes b:
[8 7]

Solução (x, y): [3.4 1.2]


Passo a Passo Data Science Academy:

- A = np.array(...): Criamos a matriz A com os coeficientes de x e y em cada equação. A primeira linha [2, 1] corresponde a 2x + 1y. A segunda linha [1, 3] corresponde a 1x + 3y.

- b = np.array(...): Criamos o vetor b com os resultados de cada equação.

- x = np.linalg.solve(A, b): Esta é a maneira mais eficiente e numericamente estável de resolver um sistema de equações lineares no NumPy. Ela encontra o vetor x que satisfaz a equação A @ x = b. O primeiro elemento do resultado é x e o segundo é y.

#### Exercício 10: Extraindo a Borda de Uma Matriz (Nível Ninja Pro Master das Galáxias Plus)

Dada uma matriz 5x5, crie um novo array 1D contendo todos os elementos da borda da matriz, em sentido horário, começando pelo canto superior esquerdo.

In [28]:
# Cria a matriz
matriz = np.arange(25).reshape(5, 5)

In [29]:
print(matriz)

[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]
 [15 16 17 18 19]
 [20 21 22 23 24]]


In [30]:
# Solução

# 1. Extrair a linha de cima
linha_topo = matriz[0, :]

# 2. Extrair a coluna da direita (sem o elemento do topo que já pegamos)
coluna_direita = matriz[1:, -1]

# 3. Extrair a linha de baixo em ordem reversa (sem o elemento da direita)
linha_baixo = matriz[-1, :-1][::-1]

# 4. Extrair a coluna da esquerda em ordem reversa (sem os elementos dos cantos)
coluna_esquerda = matriz[1:-1, 0][::-1]

# 5. Concatenar todas as partes
borda = np.concatenate([linha_topo, coluna_direita, linha_baixo, coluna_esquerda])

print("Matriz Original:")
print(matriz)
print("\nBorda da Matriz:")
print(borda)

# Saída esperada: [ 0  1  2  3  4  9 14 19 24 23 22 21 20 15 10  5]

Matriz Original:
[[ 0  1  2  3  4]
 [ 5  6  7  8  9]
 [10 11 12 13 14]
 [15 16 17 18 19]
 [20 21 22 23 24]]

Borda da Matriz:
[ 0  1  2  3  4  9 14 19 24 23 22 21 20 15 10  5]


Passo a Passo Data Science Academy:

- linha_topo = matriz[0, :]: Seleciona a primeira linha inteira.

- coluna_direita = matriz[1:, -1]: Seleciona a última coluna (-1), a partir da segunda linha (1:), para evitar repetir o elemento do canto.

- linha_baixo = matriz[-1, :-1][::-1]: Seleciona a última linha (-1), até a penúltima coluna (:-1), e depois inverte a ordem com [::-1].
<!-- Trabalho Desenvolvido na Data Science Academy - www.datascienceacademy.com.br -->
- coluna_esquerda = matriz[1:-1, 0][::-1]: Seleciona a primeira coluna (0), da segunda (1:) até a penúltima (-1) linha, e inverte a ordem.

- np.concatenate([...]): Junta todos os arrays 1D em um único array.

# Fim